### PR TITLE
fix(arc): Per-CID locking. Map CID to lock.

### PR DESCRIPTION
**Pull request by frrist**

### CODE REVIEW

File: arc_cache.go [hunk 1]
```diff
@@ -28,28 +33,76 @@ func newARCCachedBS(ctx context.Context, bs Blockstore, lruSize int) (*arccache,
...
+
+	// lock the map of cid->locks.
+	b.arcLksMu.Lock()
+	defer b.arcLksMu.Unlock()
+
+	// check if we have a lock for this content.
+	lk, hasLk := b.arcLks[k]
+	if has && hasLk {
+		// cache and lock hit.
+		lk.Lock()
```
Comment (gammazero) [inline arc_cache.go:85]: If another goroutine has already locked this CID, then locking here will wait to acquire the lock for the CID.  However, `b.arcLksMu` is not unlocked until this function exits.  This means that no other CID can be locked if there is more than one lock on the same CID.

Comment (frrist) [reply arc_cache.go:85]: oof, thanks for catching this. I'll swap this out for an RWMutex in a subsequent commit.

Comment (frrist) [reply arc_cache.go:85]: As an alternative to this, I've filed https://github.com/ipfs/go-ipfs-blockstore/pull/68 that uses a `sync.Map`.

File: arc_cache.go [hunk 2]
```diff
@@ -141,9 +220,11 @@ func (b *arccache) PutMany(bs []blocks.Block) error {
 	for _, block := range bs {
 		// call put on block if result is inconclusive or we are sure that
 		// the block isn't in storage
-		if has, _, ok := b.hasCached(block.Cid()); !ok || (ok && !has) {
+		has, _, ok, release := b.hasCachedSync(block.Cid())
+		if !ok || (ok && !has) {
 			good = append(good, block)
 		}
+		defer release()
```
Comment (frrist) [inline arc_cache.go:227]: Exploring alternatives to this as looping on defer is :vomiting_face:

**Comment by Stebalien** [comment]
Closing based on https://github.com/ipfs/go-ipfs-blockstore/issues/64#issuecomment-829246042.

**Pull request closed without merging**